web/src/people.ts

```diff
@@ -1814,17 +1815,35 @@ export function is_displayable_conversation_participant(user_id: number): boolea
     return !is_valid_bot_user(user_id) && is_person_active(user_id);
 }
 
-export function initialize(my_user_id: number, params: StateData["people"]): void {
-    for (const person of params.realm_users) {
-        add_active_user(person);
-    }
-
-    for (const person of params.realm_non_active_users) {
-        non_active_user_dict.set(person.user_id, person);
-        _add_user(person);
-    }
+export async function initialize(
+    my_user_id: number,
+    people_params: StateData["people"],
+    user_group_params: StateData["user_groups"],
+): Promise<void> {
+    user_data.initialize(user_group_params);
+    // We directly use the valid user IDs from the user_data module and
+    // fetch user data for them. Later, we will only fetch them on need basis.
+    const valid_user_ids = [...user_data.get_valid_user_ids()];
+    await user_data.get(valid_user_ids).then(
+        (users) => {
+            for (const user of users) {
+                if (user.is_active) {
+                    add_active_user(user);
+                } else {
+                    non_active_user_dict.set(user.user_id, user);
+                    _add_user(user);
+                }
+            }
+        },
+        (error: unknown) => {
+            if (error instanceof Error) {
+                blueslip.error(error.message);
+            }
+            blueslip.error("Failed to fetch user data");
+        },
+    );
```

Comment:
I think we only want to do this fetch if we don't have all the users in `valid_user_ids` inside `people_params`.

Reply:
Changed to do that.